plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -234,6 +263,106 @@ public void onFailure(Exception e) {
+            } else {
+                String errorMessage = formatAgentErrorMessage(searchHits);
+
+                actionListener.onFailure(new OpenSearchStatusException(errorMessage, RestStatus.CONFLICT));
+            }
+
+        }, e -> {
+            if (e instanceof IndexNotFoundException) {
+                actionListener.onResponse(true);
+                return;
+            }
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+    }
+
+    private void checkIngestPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        checkPipelineBeforeDeleteModel(modelId, actionListener, "ingest", GetPipelineRequest::new, GetPipelineAction.INSTANCE);
+
+    }
+
+    private void checkSearchPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        checkPipelineBeforeDeleteModel(modelId, actionListener, "search", GetSearchPipelineRequest::new, GetSearchPipelineAction.INSTANCE);
+
+    }
+
+    private void checkPipelineBeforeDeleteModel(
+        String modelId,
+        ActionListener<Boolean> actionListener,
+        String pipelineType,
+        Supplier<ActionRequest> requestSupplier,
+        ActionType actionType
+    ) {
+        ActionRequest request = requestSupplier.get();
+        client.execute(actionType, request, ActionListener.wrap(pipelineResponse -> {
+            Map<String, Object> allConfigMap = XContentHelper.convertToMap(JsonXContent.jsonXContent, pipelineResponse.toString(), true);
+            List<String> allDependentPipelineIds = findDependentPipelinesEasy(allConfigMap, modelId);
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            String
+                                .format(
+                                    Locale.ROOT,
+                                    "%d %s pipelines are still using this model, please delete or update the pipelines first: %s",
+                                    allDependentPipelineIds.size(),
+                                    pipelineType,
+                                    Arrays.toString(allDependentPipelineIds.toArray(new String[0]))
+                                ),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkDownstreamTaskBeforeDeleteModel(String modelId, Boolean isHidden, ActionListener<DeleteResponse> actionListener) {
+        // Now 3 is: 1. Agent 2. Search pipeline 3. ingest pipeline
+        CountDownLatch countDownLatch = new CountDownLatch(3);
```

Comment:
Do we need to block the thread?

Can we use `CompletableFuture` instead of `CountDownLatch` to unblock the thread? 

A code example could be:

```
private void checkDownstreamTaskBeforeDeleteModel(String modelId, Boolean isHidden, ActionListener<DeleteResponse> actionListener) {
    AtomicBoolean failed = new AtomicBoolean(false);
    ConcurrentLinkedQueue<String> errorMessages = new ConcurrentLinkedQueue<>();

    CompletableFuture<Void> allChecks = CompletableFuture.allOf(
        runCheck(modelId, this::checkAgentBeforeDeleteModel, failed, errorMessages),
        runCheck(modelId, this::checkIngestPipelineBeforeDeleteModel, failed, errorMessages),
        runCheck(modelId, this::checkSearchPipelineBeforeDeleteModel, failed, errorMessages)
    );

    allChecks.thenRun(() -> {
        if (failed.get()) {
            actionListener.onFailure(new OpenSearchStatusException(String.join(". ", errorMessages), RestStatus.CONFLICT));
        } else {
            deleteModel(modelId, isHidden, actionListener);
        }
    });
}

private CompletableFuture<Void> runCheck(String modelId, BiConsumer<String, ActionListener<Boolean>> checkMethod, 
                                         AtomicBoolean failed, ConcurrentLinkedQueue<String> errorMessages) {
    CompletableFuture<Void> future = new CompletableFuture<>();
    if (failed.get()) {
        future.complete(null);
        return future;
    }

    checkMethod.accept(modelId, ActionListener.wrap(
        success -> {
            if (!success) {
                failed.set(true);
                errorMessages.add("Check failed for " + checkMethod.getClass().getSimpleName());
            }
            future.complete(null);
        },
        error -> {
            failed.set(true);
            errorMessages.add(error.getMessage());
            future.complete(null);
        }
    ));

    return future;
}
```

Warning: I didn't test this code though :)

Reply:
Since we use Actionlisener here, CountDownLatch won't block the thread and we don't need to await. And since we don't contact between each downstream thread, CountDownLatch is already enough here.

Reply:
I prefer `CompletableFuture` which is more cleaner, but upto you. 

In addition, can we fail fast, rather than waiting? If we see a model is being used in Agent, we don't need to wait to get result from Search or Index pipeline, right?

Reply:
It's a trade off. In current way, although we wait for all three downstream checking finishing, but in one call customer could collect messages from all downstream task, so he doesn't need to call it more times.

Reply:
If there is a use case where customers want to know about all the resources where a model is being used, we can build a separate API for that purpose.

Please consider failing fast to reduce the latency of the delete API. The primary goal of the delete API is to delete the resource with minimal latency, not to inform customers about which resources are using the model (that is just a byproduct). Therefore, I don’t think there’s any question about trade-offs.

Reply:
Got it. Now once one search process fails, it will return the result.